[AArch64] Support symmetric complex deinterleaving with higher factors

david-arm opened the pull request:
For loops such as this:

```
struct foo {
  double a, b;
};

void foo(struct foo *dst, struct foo *src, int n) {
  for (int i = 0; i < n; i++) {
    dst[i].a += src[i].a * 3.2;
    dst[i].b += src[i].b * 3.2;
  }
}
```

the complex deinterleaving pass will spot that the deinterleaving
associated with the structured loads cancels out the interleaving
associated with the structured stores. This happens even though
they are not truly "complex" numbers because the pass can handle
symmetric operations too. This is great because it means we can
then perform normal loads and stores instead. However, we can also
do the same for higher interleave factors, e.g. 4:

```
struct foo {
  double a, b, c, d;
};

void foo(struct foo *dst, struct foo *src, int n) {
  for (int i = 0; i < n; i++) {
    dst[i].a += src[i].a * 3.2;
    dst[i].b += src[i].b * 3.2;
    dst[i].c += src[i].c * 3.2;
    dst[i].d += src[i].d * 3.2;
  }
}
```

This PR extends the pass to effectively treat such structures as
a set of complex numbers, i.e.

```
struct foo_alt {
  std::complex<double> x, y;
};
```

with equivalence between members:

```
  foo_alt.x.real == foo.a
  foo_alt.x.imag == foo.b
  foo_alt.y.real == foo.c
  foo_alt.y.imag == foo.d
```

I've written the code to handle sets with arbitrary numbers of
complex values, but since we only support interleave factors
between 2 and 4 I've restricted the sets to 1 or 2 complex
numbers. Also, for now I've restricted support for interleave
factors of 4 to purely symmetric operations only. However, it
could also be extended to handle complex multiplications,
reductions, etc.

Fixes: https://github.com/llvm/llvm-project/issues/144795

igogo-x86 commented on llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:
```diff
@@ -67,6 +67,7 @@
 #include "llvm/CodeGen/TargetLowering.h"
 #include "llvm/CodeGen/TargetSubtargetInfo.h"
 #include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Intrinsics.h"
```
I think we don't need that

igogo-x86 replied:
Ah, yes, you are right!

igogo-x86 commented on llvm/test/CodeGen/AArch64/complex-deinterleaving-symmetric-fixed.ll:
```diff
@@ -0,0 +1,76 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
I suppose this patch doesn't affect fixed-width? This test shows the same output before and after patch

david-arm replied:
Yeah so the pass already handles `@simple_symmetric_muladd2`, but I didn't add support for the fixed-width patterns in `@simple_symmetric_muladd4` mainly because I'd expect the VectorCombine pass to have already removed the shuffles before we reach the complex deinterleaving pass. However, if we find cases where actually it would be useful to support fixed-width vectors it should be easy enough to add.

igogo-x86 commented on llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:
```diff
@@ -2033,24 +2190,38 @@ ComplexDeinterleavingGraph::identifySplat(Value *R, Value *I) {
+  // The splats must meet the following requirements:
+  //   1. Must either be all instructions or all values.
+  //   2. Non-constant splats must live in the same block.
+  auto FirstValAsInstruction = dyn_cast<Instruction>(Vals[0].Real);
```
I feel like it would look nice if we split here into two branches and handle them separately:
```
if (FirstIsInst) {
  // Check that all are instructions and have the same parent
} else {
  // Just check that all Vals are not instructions
}
```

igogo-x86 commented on llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:
```diff
@@ -305,10 +351,26 @@ class ComplexDeinterleavingGraph {
     CompositeNodes.push_back(Node);
-    if (Node->Real)
-      CachedResult[{Node->Real, Node->Imag}] = Node;
+    if (Node->Vals[0].Real) {
```
Braces on a simple single-statement body

david-arm committed:
Add tests

david-arm committed:
[AArch64] Support symmetric complex deinterleaving with higher factors

For loops such as this:

struct foo {
  double a, b;
};

void foo(struct foo *dst, struct foo *src, int n) {
  for (int i = 0; i < n; i++) {
    dst[i].a += src[i].a * 3.2;
    dst[i].b += src[i].b * 3.2;
  }
}

the complex deinterleaving pass will spot that the deinterleaving
associated with the structured loads cancels out the interleaving
associated with the structured stores. This happens even though
they are not truly "complex" numbers because the pass can handle
symmetric operations too. This is great because it means we can
then perform normal loads and stores instead. However, we can also
do the same for higher interleave factors, e.g. 4:

struct foo {
  double a, b, c, d;
};

void foo(struct foo *dst, struct foo *src, int n) {
  for (int i = 0; i < n; i++) {
    dst[i].a += src[i].a * 3.2;
    dst[i].b += src[i].b * 3.2;
    dst[i].c += src[i].c * 3.2;
    dst[i].d += src[i].d * 3.2;
  }
}

This PR extends the pass to effectively treat such structures as
a set of complex numbers, i.e.

struct foo_alt {
  std::complex<double> x, y;
};

with equivalence between members:

  foo_alt.x.real == foo.a
  foo_alt.x.imag == foo.b
  foo_alt.y.real == foo.c
  foo_alt.y.imag == foo.d

I've written the code to handle sets with arbitrary numbers of
complex values, but since we only support interleave factors
between 2 and 4 I've restricted the sets to 1 or 2 complex
numbers. Also, for now I've restricted support for interleave
factors of 4 to purely symmetric operations only. However, it
could also be extended to handle complex multiplications,
reductions, etc.

Fixes: https://github.com/llvm/llvm-project/issues/144795

igogo-x86 reviewed (approved):
LGTM!

david-arm committed:
Address review comment
